### ISSUE TITLE
bug in parse_css.cpp line 188 and 189

### ISSUE THREAD

**Issue by kevinhendricks** [csstidy-c/csstidy#5]
In parse_css.cpp line 188 and 189 the case of cur_selector being empty is not handled:
```
				int lastpos = cur_selector.length()-1;
				if(!( (ctype_space(cur_selector[lastpos]) || is_token(cur_selector,lastpos) && cur_selector[lastpos] == ',') && ctype_space(css_input[i])))
				{
					cur_selector += css_input[i];
				}
```

The if condition in line 189 needs to handle the case of lastpos being -1 here.

So it needs to add:

 (lastpos == -1) || 

to the front of condition being tested in line 189 to parallel similar code in parse_css.cpp used in lines 99 and 100.

**Comment by csstidy-c** [comment]
fixed with 299e9c35132c4c152c5425443a1157a6598addd2
thank you!

**Comment by kevinhendricks** [comment]
FYI,
Sigil, a GPL3 epub editor program,  (https://github.com/Sigil-Ebook/Sigil)

has extracted and modified csstidy parser code to create a standalone css parser (no optimization) that we are then porting to Qt for internal use in Sigil.

During the porting to Qt, Qt's debug asserts detected all 3 of these error cases, so I thought I would pass them along.

If you want to see the restructured code you can check out this post in our user discussion forum:

https://www.mobileread.com/forums/showpost.php?p=4080944&postcount=64

I did not see your personal copyright notice anywhere in your existing code so I would be happy to add it in appropriate cases in our code.  Just let me know.

**Comment by csstidy-c** [comment]
haha... great. I already clicked on your account and then on Sigil and checked some commits (also by you). Hence I already know what Sigil is. :-)
Thanks for the details. I'll later take a closer look.
The copyright notice of Floele was not modified by me until now but I listed all authors in AUTHORS file - including you.
:-)

**Comment by kevinhendricks** [comment]
Do you want me to add all authors from that file to the top of each of the 3 csstidy derived files now in Sigil?

I would be happy to do so since the authors file itself is not included into Sigil.  Something like:

CSSTidy Portions Copyright:

and then the list of names.  Right now Sigil is using the following file header:

```
/************************************************************************
 **
 **  Copyright (C) 2021  Kevin B. Hendricks, Stratford, Ontario, Canada
 **
 **  This file is part of Sigil.
 **
 **  Sigil is free software: you can redistribute it and/or modify
 **  it under the terms of the GNU General Public License as published by
 **  the Free Software Foundation, either version 3 of the License, or
 **  (at your option) any later version.
 **
 **  Sigil is distributed in the hope that it will be useful,
 **  but WITHOUT ANY WARRANTY; without even the implied warranty of
 **  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 **  GNU General Public License for more details.
 **
 **  You should have received a copy of the GNU General Public License
 **  along with Sigil.  If not, see <http://www.gnu.org/licenses/>.
 **
 ** Extracted and modified from:
 ** CSSTidy Copyright 2005-2007 Florian Schmitz
 ** Available under the LGPL 2.1
 ** You should have received a copy of the GNU Lesser General Public License
 ** along with this program.  If not, see <http://www.gnu.org/licenses/>.
 **
 *************************************************************************/
```

**Comment by csstidy-c** [comment]
Yes, please use "portions" to add all people from AUTHORS file. That would be great.

**Comment by kevinhendricks** [comment]
Done:  see commit: https://github.com/Sigil-Ebook/Sigil/commit/4b8fa410e5d0b2ba642ea7bc0a68e528ad7f8d06

which among other things modifies the above to end with:

```
 ** Extracted and modified from:
 ** CSSTidy (https://github.com/csstidy-c/csstidy)
 **
 ** CSSTidy Portions Copyright:
 **   Florian Schmitz <floele@gmail.com>
 **   Thierry Charbonnel
 **   Will Mitchell <aethon@gmail.com>
 **   Brett Zamir <brettz9@yahoo.com>
 **   sined_ <sined_@users.sourceforge.net>
 **   Dmitry Leskov <git@dmitryleskov.com>
 **   Kevin Coyner <kcoyner@debian.org>
 **   Tuukka Pasanen <pasanen.tuukka@gmail.com>
 **   Frank W. Bergmann <csstidy-c@tuxad.com>
 **   Frank Dana <ferdnyc@gmail.com>
 **
 ** CSSTidy us Available under the LGPL 2.1
 ** You should have received a copy of the GNU Lesser General Public License
 ** along with this program.  If not, see <http://www.gnu.org/licenses/>.
```

If I run into any other bugs that exist in both versions, I will pass along the fixes.  

You of course have my permission to use any of the code I wrote for Sigil's CSSParser.cpp, CSSParser.h, CSSUtils.h, CSSUtils.cpp, CSSProperties.cpp, CSSProperties.h under any license you need for your project.

Take care.
